### PR TITLE
Implement double space full stop, switch caps lock from long click to double click

### PR DESCRIPTION
**Pull request by kfli716**

### CODE REVIEW

File: app/src/main/java/hk/eduhk/typeduck/ime/keyboard/KeyboardView.java [hunk 1]
```diff
@@ -1091,7 +1097,12 @@ private void detectAndSendKey(int index, int x, int y, long eventTime, KeyEventT
         Timber.d(
             "\t<TrimeInput>\tdetectAndSendKey()\tModifierKey, key.getEvent, KeyLabel=%s",
             key.getLabel());
-        setModifier(key);
+        if (isDoubleClick) {
```
Comment (graphemecluster) [inline app/src/main/java/hk/eduhk/typeduck/ime/keyboard/KeyboardView.java]: For safety
```suggestion
        if (key.isShift() && isDoubleClick) {
```

File: app/src/main/java/hk/eduhk/typeduck/ime/keyboard/KeyboardView.java [hunk 2]
```diff
@@ -1104,10 +1115,47 @@ private void detectAndSendKey(int index, int x, int y, long eventTime, KeyEventT
...
+             is used to track the label of the last space key pressed.
+           */
+          switch (key.getLabel()) {
+            case "Space":
+              // need full stop when english mode
+              if (isDoubleClick) {
+                // The isHalfShapePunct is used to check whether ". " or "。" should be entered after double space.
+                mKeyboardActionListener.onDoubleSpace(
+                        lastSpaceWasConfirm ? 0 : 1,
+                        mKeyboard.isHalfShapePunct() ? ". " : "。"
```
Comment (graphemecluster) [inline app/src/main/java/hk/eduhk/typeduck/ime/keyboard/KeyboardView.java]: Subtle but
```suggestion
                    lastSpaceWasConfirm ? 0 : 1,
                    mKeyboard.isHalfShapePunct() ? ". " : "。"
```

File: app/src/main/java/hk/eduhk/typeduck/ime/keyboard/KeyboardView.java [hunk 3]
```diff
@@ -1104,10 +1115,47 @@ private void detectAndSendKey(int index, int x, int y, long eventTime, KeyEventT
...
+          /* When space key is pressed, there could be two labels depending on the state.
+             If you are typing Chinese and press double space, the first space key will be labeled as Confirm
+             and the second space is labeled as Space. In this case, the Confirm space keys ejects the Chinese words
+             you are typing and there is no trailing space. So we don't want to delete anything at the end of the input.
+             If you are not typing Chinese and press double space, both of the space keys will be labeled as Space. And
+             the first space key will input a empty character, so when pressing the space key for the second time, we need
+             to delete the trailing space. And this is handled in the onDoubleSpace function. The boolean lastSpaceWasConfirm
+             is used to track the label of the last space key pressed.
+           */
+          switch (key.getLabel()) {
```
Comment (graphemecluster) [inline app/src/main/java/hk/eduhk/typeduck/ime/keyboard/KeyboardView.java]: Use `Rime.isComposing()` instead – the key labels may be localized later

File: app/src/main/java/hk/eduhk/typeduck/ime/text/TextInputManager.kt [hunk 4]
```diff
@@ -563,4 +563,9 @@ class TextInputManager private constructor() :
         trime.showDialogAboveInputView(builder.create())
     }
     */
+
+    // added to handle doubleSpaceFullStop
+    override fun onDoubleSpace(charToDel: Int, punct: CharSequence) {
+        activeEditorInstance.doubleSpaceEvent(charToDel, punct)
+    }
```
Comment (graphemecluster) [inline app/src/main/java/hk/eduhk/typeduck/ime/text/TextInputManager.kt]: I prefer moving this above `showOptionsDialog`

**Review by graphemecluster — COMMENTED** [review]
- Please remove [KeyboardView.java#L1480-L1487](https://github.com/TypeDuck-HK/TypeDuck-Android/blob/84ef8d49b93678d2199c28e9730c82e3e7f2bb88/app/src/main/java/hk/eduhk/typeduck/ime/keyboard/KeyboardView.java#L1480-L1487)
- Could you rebase your change and help me uncomment the option in [app/src/main/res/xml/preference.xml](https://github.com/TypeDuck-HK/TypeDuck-Android/blob/50ea4cb6c0f3685acce9d957e1582ca5e4843eaf/app/src/main/res/xml/preference.xml#L54-L61)?

File: app/src/main/java/hk/eduhk/typeduck/ime/keyboard/KeyboardView.java (hunk 2 above)
Comment (graphemecluster) [reply app/src/main/java/hk/eduhk/typeduck/ime/keyboard/KeyboardView.java]: I like these details 👍🏻

**Comment by graphemecluster** [comment]
Oh, but the behavior does not match what Gboard and iOS built-in keyboard currently do – the click when composing should not be counted towards double click. That means we should only detect double click on spacebar when `Rime.isComposing()` is false and to insert a full stop during composition a user needs to click the spacebar a single time for confirmation to exit the composition mode before a double click.

Plus, I plan to add and option to change the behavior of the spacebar when composing (flip page, insert candidate, move to next candidate, keep input), so spacebar double clicking should be disabled during composition to prevent any disruption.

File: app/src/main/java/hk/eduhk/typeduck/ime/keyboard/KeyboardView.java [hunk 5]
```diff
@@ -1091,7 +1097,12 @@ private void detectAndSendKey(int index, int x, int y, long eventTime, KeyEventT
         Timber.d(
             "\t<TrimeInput>\tdetectAndSendKey()\tModifierKey, key.getEvent, KeyLabel=%s",
             key.getLabel());
-        setModifier(key);
+        if (isDoubleClick) {
+          mKeyboard.setShifted(true, true);
+          //isDoubleClick = false;
```
Comment (graphemecluster) [inline app/src/main/java/hk/eduhk/typeduck/ime/keyboard/KeyboardView.java:1102]: Do you think this line is needed? Let’s remove it if not

**Comment by graphemecluster** [comment]
But remember, shift double clicking should still work as usual during composition

File: app/src/main/java/hk/eduhk/typeduck/ime/keyboard/KeyboardView.java (hunk 5 above)
Comment (graphemecluster) [inline app/src/main/java/hk/eduhk/typeduck/ime/keyboard/KeyboardView.java]: Do you think this line is needed? Let’s remove it if not

**Comment by kfli716** [comment]
Changes:
1. Changed the code accoding to the suggestions.
2. Added a function isSpace() to Key.java to distinguish if the pressed key is the space key, without using key label.
3. Changed the behavior to be more aligned with the GBoard. Now the doubleSpaceToFullStop function will not trigger if the last character is a space key or in " ！？、，。!?.,". And if it triggers, the full stop generated is determined by the input language being used.

**Comment by graphemecluster** [comment]
Done? If so, I’m going to investigate this today

**Comment by graphemecluster** [comment]
Never mind, tomorrow is fine

**Comment by kfli716** [comment]
Yes, and if you have time, please also review the PR about launch issue (the CMake file) and let me know whether there is any problem, thank you very much! 🙏

**Pull request merged**